subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/result/ComponentSelectorSerializer.java

```diff
@@ -121,6 +123,8 @@ private void writeCapabilities(Encoder encoder, Collection<Capability> capabilit
     }
 
     public void write(Encoder encoder, ComponentSelector value) throws IOException {
+        attributeContainerSerializer.reset();
```

Comment:
I wonder if that's possible and working across all serializers. Should it be an assertion instead that we only use stateful serializers only ever once for writing?

Reply:
Write should really always be called once, so this line is pretty much redundant, but I added it for "symmetry".